File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/bannerview/BannerCallback.kt
package com.cleveradssolutions.plugin.flutter.bannerview

import com.cleveradssolutions.plugin.flutter.bridge.base.MappedCallback
import com.cleveradssolutions.plugin.flutter.bridge.base.MappedMethodHandler
import com.cleveradssolutions.plugin.flutter.util.toMap
import com.cleversolutions.ads.AdError
import com.cleversolutions.ads.AdImpression
import com.cleversolutions.ads.AdViewListener
import com.cleversolutions.ads.android.CASBannerView

class BannerCallback(
    private val sizeListener: BannerSizeListener,
    handler: MappedMethodHandler<*>,
    id: String
) : MappedCallback(handler, id), AdViewListener {

    override fun onAdViewLoaded(view: CASBannerView) {
        invokeMethod("onAdViewLoaded")

        val size = view.size
        sizeListener.updateSize(size.width, size.height)
    }

    override fun onAdViewFailed(view: CASBannerView, error: AdError) {
        invokeMethod("onAdViewFailed", "error" to error.message)

        sizeListener.updateSize(0, 0)
    }

    override fun onAdViewPresented(view: CASBannerView, info: AdImpression) {
        invokeMethod("onAdViewPresented", info.toMap())
    }

    override fun onAdViewClicked(view: CASBannerView) {
        invokeMethod("onAdViewClicked")
    }

}

File: ios/Classes/BannerView/BannerCallback.swift
//
//  AppOpenMethodHandler.swift
//  clever_ads_solutions
//
//  Copyright © 2024 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import CleverAdsSolutions

class BannerCallback: MappedCallback, CASBannerDelegate {
    private let sizeListener: BannerSizeListener

    init(_ sizeListener: BannerSizeListener, _ handler: AnyMappedMethodHandler, _ id: String) {
        self.sizeListener = sizeListener
        super.init(handler, id)
    }

    func bannerAdViewDidLoad(_ view: CASBannerView) {
        invokeMethod("onAdViewLoaded")

        let size = view.intrinsicContentSize
        sizeListener.updateSize(size.width, size.height)
    }

    func bannerAdView(_ adView: CASBannerView, didFailWith error: CASError) {
        invokeMethod("onAdViewFailed", ["error": error.message])

        sizeListener.updateSize(0, 0)
    }

    func bannerAdView(_ adView: CASBannerView, willPresent impression: CASImpression) {
        invokeMethod("onAdViewPresented", impression.toDict())
    }

    func bannerAdViewDidRecordClick(_ adView: CASBannerView) {
        invokeMethod("onAdViewClicked")
    }
}
